compiler/src/dotty/tools/dotc/typer/Implicits.scala

```diff
@@ -928,9 +933,70 @@ trait Implicits:
           if normalImports.isEmpty then suggestedImports else normalImports
         end hiddenImplicitsAddendum
 
-        msg(userDefined.getOrElse(
-          em"no implicit argument of type $pt was found${location("for")}"))() ++
-        hiddenImplicitsAddendum
+
+        /** Extracting the message from a method parameter, e.g. in
+         *
+         *  trait Foo
+         *
+         *  def foo(implicit @annotation.implicitNotFound foo: Foo)
+         */
+        def userDefinedParamMessage = paramSymWithMethodTree.flatMap { (sym, tree) =>
+          userDefinedMsg(sym, defn.ImplicitNotFoundAnnot).map { rawMsg =>
+            val paramSyms = typeVariablesInScope(sym)
+            val paramTypeRefs = paramSyms.map(_.typeRef.typeConstructor)
+
+            val argTypes = tree match {
+              case TypeApply(Select(qual, _), targs) =>
+                val methodOwnerType = qual.tpe
+                val methodOwner = tree.symbol.owner
+                val methodTypeParams = tree.symbol.paramSymss.flatten.filter(_.isType)
+                val methodTypeArgs = targs.map(_.tpe)
+                paramTypeRefs.map(_.asSeenFrom(methodOwnerType, methodOwner).subst(methodTypeParams, methodTypeArgs))
+              case _ => paramTypeRefs
```

Comment:
`asSeenFrom` is also needed when the tree isn't a polymorphic method:
```scala
import scala.annotation.implicitNotFound

class Foo[A] {
  def foo(implicit @implicitNotFound("A = ${A}") x: String): Unit = ???
}

object Test {
  val x: Foo[Int] = ???
  x.foo
}
```
Currently outputs `A = A` instead of `A = Int`